Tasks API - Seek/Keyset based pagination (#115)
meilisearch/specifications: text/0060-tasks-api.md

}
```

#### 9. Paginate `task` resource lists

The API endpoints `/tasks` and `indexes/{indexUid}/tasks` are browsable using a keyset based pagination.

##### 9.1 Why a keyset based pagination?

Keyset-based pagination is more appropriate when the data can grow or shrink quickly in terms of magnitude.

###### 9.1.1 Pros

The performance is better than the not-so-good but old pagination with `offset`/`limit`.

> Cursor pagination keeps the results consistent between each page as the data evolves. It avoids the [Page Drift effect](https://use-the-index-luke.com/sql/partial-results/fetch-next-page), especially when the data is sorted from the most recent to the oldest.

Reviewer: Is cursor still the term you want to use here?

Author: No! Good spot!

Moreover, the performance is superior to traditional pagination since the computational complexity remains constant to reach the identifier marking the beginning of the new slice to be returned from a hash table.

###### 9.1.2 Cons

The main drawback of this type of pagination is that it does not navigate within a finite number of pages. It is also limited to a precise sorting criterion on unique identifiers ordered sequentially.

##### 9.2 Response attributes

| field | type | description                          |
|-------|------|--------------------------------------|
| limit | integer  | Default `20`. |

> | after | integer - nullable  | Represents the query parameter to send to fetch the next slice of the results. The first item for the next slice starts at `after+1`. When the returned value is null, it means that all the data have been browsed in the given order. |

Reviewer: ```suggestion
| after | integer - nullable  | Used to fetch the next slice of the results. The first item for the next slice starts at `after+1`. When the returned value is null, it means that all the data have been browsed in the given order. |
```

##### 9.3 GET query parameters

| field | type | required | description  |
|-------|------|----------|--------------|

> | limit | integer  | No       | Default `20`. Limit on the number of tasks to be returned. |
> | after | integer  | No       | Limit results to tasks with uids greater/lower than the specified uid. |

Reviewer 2: ```suggestion
| limit | integer  | No       | Default `20`. Limit on the number of tasks to be returned. |
| from | integer  | No       | Limit results to tasks with uids equal to and lower than this uid. |
```

---

**Initial default slice of `tasks`**

`GET` - `/tasks`

```json
{
    "results": [
        {
            "uid": 1350,
            "indexUid": "movies",
            "type": "documentAddition",
            ...,
        },
        ...,
        {
            "uid": 1330,
            "indexUid": "movies_reviews",
            "type": "documentAddition",
            ...,
        }
    ],
    "limit": 20,
    "after": 1330
}
```

**Request the next slice of `tasks` items with a limit of `50` tasks**

`GET` - `/tasks?after=1330&limit=50`

```json
{
    "results": [
        {
            "uid": 1329,
            "indexUid": "movies",
            "type": "documentAddition",
            ...,
        },
        ...,
        {
            "uid": 1279,
            "indexUid": "movies",
            "type": "settingsUpdate",
            ...,
        }
    ],
    "limit": 50,
    "after": 1279
}
```

**End of cursor pagination**

> `GET` - `/tasks?after=20`

Reviewer 2: ```suggestion
`GET` - `/tasks?from=20`
```

```json
{
    "results": [
        {
            "uid": 19,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        },
        ...,
        {
            "uid": 0,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        }
    ],
    "limit": 10,
    "after": null
}
```

> - 💡 `after` response parameter is null because there are no more `tasks` to fetch. It means that the response represents the last slice of results for the given resource list.

Reviewer 2: ```suggestion
- 💡 `next` response parameter is null because there are no more `tasks` to fetch. It means that the response represents the last slice of results for the given resource list.
```

> ##### 9.5 Behaviors for `limit` and `after` query parameters
> 
> ###### 9.5.1 `limit`
> 
> - If `limit` is not set, the default value is chosen.
> - If `limit` is not an integer, the default value is chosen.
> 
> ###### 9.5.2 `after`
> 
> - If `after` is set with an out of bounds task `uid`, the response returns an empty `results` array and `after` is set to `null`.

Reviewer 2: ```suggestion
##### 9.5 Behaviors for `limit` and `from` query parameters

###### 9.5.1 `limit`

- If `limit` is not set, the default value is chosen.

###### 9.5.2 `from`

- If `from` is set with an out of bounds task `uid`, the response returns the tasks that are the nearest to the specified uid, the `next` field is set to the next page. It will be equivalent to call the `/tasks` route without any parameter.
```

`GET` - `/tasks?after=1330&limit=50`

```json
{
    "results": [
        {
            "uid": 1329,
            "indexUid": "movies",
            "type": "documentAddition",
            ...,
        },
        ...,
        {
            "uid": 1279,
            "indexUid": "movies",
            "type": "settingsUpdate",
            ...,
        }
    ],
    "limit": 50,
    "after": 1279
}
```

**End of cursor pagination**

`GET` - `/tasks?after=20`

```json
{
    "results": [
        {
            "uid": 19,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        },
        ...,
        {
            "uid": 0,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        }
    ],
    "limit": 10,
    "after": null
}
```

- 💡 `after` response parameter is null because there are no more `tasks` to fetch. It means that the response represents the last slice of results for the given resource list.

##### 9.5 Behaviors for `limit` and `after` query parameters

###### 9.5.1 `limit`

- If `limit` is not set, the default value is chosen.

> - If `limit` is set and it is not between the minimum and maximum values, the default value is chosen.

Reviewer 3: What are the min and max values? They are hardcoded in the engine?

Author: min is always 0, while maximum is the total number of task. I will rephrase it! Thanks @brunoocasali

Author: I think the `maximum` could be adjusted given some tests!

}
```

#### 9. Paginate `task` resource lists

> The API endpoints `/tasks` and `indexes/{indexUid}/tasks` are browsable using a keyset based pagination.

Reviewer 4: ```suggestion
The API endpoint `/tasks` is browsable using a keyset-based pagination.
```

##### 9.1 Why a Seek/Keyset based pagination?

Keyset-based pagination is more appropriate when the data can grow or shrink quickly in terms of magnitude.

###### 9.1.1 Pros

The performance is better than the not-so-good but old pagination with `offset`/`limit`.

Seek/Keyset pagination keeps the results consistent between each page as the data evolves. It avoids the [Page Drift effect](https://use-the-index-luke.com/sql/partial-results/fetch-next-page), especially when the data is sorted from the most recent to the oldest.

Moreover, the performance is superior to traditional pagination since the computational complexity remains constant to reach the identifier marking the beginning of the new slice to be returned from a hash table.

###### 9.1.2 Cons

> The main drawback of this type of pagination is that it does not navigate within a finite number of pages. It is also limited to a precise sorting criterion on unique identifiers ordered sequentially.

Reviewer: What do you mean with `does not navigate within a finite number of pages` ?

Author: It is not possible to access a specific page like with `offset`/`limit`, the navigation is done as you go with `from`

##### 9.2 Response attributes

| field | type | description                          |
|-------|------|--------------------------------------|
| limit | integer  | Default `20`. |
| from | integer | The first task uid returned |

> | next | integer - nullable  | Represents the value to send in `from` to fetch the next slice of the results. The first item for the next slice starts at this exact number. When the returned value is null, it means that all the data have been browsed in the given order. |

Reviewer: ```suggestion
| next | integer - nullable  | Represents the value to send in `from` to fetch the next slice of the results. The first item for the next slice starts at this exact number. When the returned value is null, it means that all the data have been browsed in the given order. `next = form +/- limit` depending on the order of the tasks |
```
For me it is a bit more clear but feel free to ignore

##### 9.3 GET query parameters

| field | type | required | description  |
|-------|------|----------|--------------|
| limit | integer  | No       | Default `20`. Limit on the number of tasks to be returned. |
| from | integer  | No       | Limit results to tasks with uids equal to and lower than this uid. |

##### 9.4 Usage examples

> This part demonstrates keyset paging in action on `/tasks`, but it should be equivalent for `indexes/:uid/tasks`. Although the uids can be "holey" on the `/indexes/:uid/tasks` endpoint if several indexes are managed within the instance. The items uid remains sorted sequentially and can be used to navigate a list of `tasks` objects.

Reviewer: What do you mean with holey?

}
```

**End of seek/keyset pagination**

`GET` - `/tasks?from=20`

```json
{
    "results": [
        {
            "uid": 19,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        },
        ...,
        {
            "uid": 0,
            "indexUid": "movies",
            "type": "documentsAddition",
            ...,
        }
    ],
    "limit": 20,
    "from": 20,
    "next": null
}
```

- 💡 `next` response parameter is null because there are no more `tasks` to fetch. It means that the response represents the last slice of results for the given resource list.

##### 9.5 Behaviors for `limit` and `from` query parameters

###### 9.5.1 `limit`

- If `limit` is not set, the default value is chosen.

###### 9.5.2 `from`

- If `from` is set with an out of bounds task `uid`, the response returns the tasks that are the nearest to the specified uid, the `next` field is set to the next page. It will be equivalent to call the `/tasks` route without any parameter.

###### 9.5.3 Errors

- 🔴 Sending a value with a different type than `Integer` for `limit` returns a [bad_request](0061-error-format-and-definitions.md#bad_request) error.

> - 🔴 Sending a value with a different type than `Integer` for `from` returns a [bad_request](0061-error-format-and-definitions.md#bad_request) error.

Reviewer: Does the bad_request specify which field is wrong?

Reviewer: Thanks for the information :)
